b2/progress.py

```diff
@@ -228,3 +229,60 @@ def write(self, data):
     def _update(self, delta):
         self.bytes_completed += delta
         self.progress_listener.bytes_completed(self.bytes_completed + self.offset)
+
+
+class StreamWithHash(object):
+    """
+    Wraps a file-like object, calculates SHA1 while reading
+    and appends hash at the end
+    """
+
+    def __init__(self, stream):
+        """
+        :param stream: the stream to read from
+        :return: None
+        """
+        self.stream = stream
+        self.digest = hashlib.sha1()
+        self.hash = None
+        self.hash_read = 0
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        return self.stream.__exit__(exc_type, exc_val, exc_tb)
+
+    def seek(self, pos):
+        assert pos == 0
+        self.stream.seek(0)
+        self.digest = hashlib.sha1()
+        self.hash = None
+        self.hash_read = 0
+
+    def read(self, size=None):
+        data = b''
+        if self.hash is None:
+            # Read some bytes from stream
+            if size is None:
+                data = self.stream.read()
+            else:
+                data = self.stream.read(size)
+
+            # Update hash
+            self.digest.update(data)
+
+            # Check for end of stream
+            if size is None or (len(data) == 0 and size != 0):
```

Comment:
The doc for file objects says:

Read at most size bytes from the file (less if the read hits EOF before obtaining size bytes). If the size argument is negative or omitted, read all data until EOF is reached. The bytes are returned as a string object. An empty string is returned when EOF is encountered immediately. (For certain files, like ttys, it makes sense to continue reading after an EOF is hit.) **Note that this method may call the underlying C function fread() more than once in an effort to acquire as close to size bytes as possible.** Also note that when in non-blocking mode, less data than was requested may be returned, even if no size parameter was given.